models/uniswap/arbitrum/uniswap_v3_arbitrum_trades.sql

```diff
@@ -0,0 +1,110 @@
+    )
+}}
+
+{% set project_start_date = '2021-06-01' %}
+
+WITH dexs AS
+(
+    --Uniswap v3
+    SELECT
+        t.evt_block_time AS block_time
+        ,t.recipient AS taker
+        ,'' AS maker
+        ,CASE WHEN amount0 < '0' THEN abs(amount0) ELSE abs(amount1) END AS token_bought_amount_raw -- when amount0 is negative it means trader_a is buying token0 from the pool
+        ,CASE WHEN amount0 < '0' THEN abs(amount1) ELSE abs(amount0) END AS token_sold_amount_raw
+        ,NULL AS amount_usd
+        ,CASE WHEN amount0 < '0' THEN f.token0 ELSE f.token1 END AS token_bought_address
+        ,CASE WHEN amount0 < '0' THEN f.token1 ELSE f.token0 END AS token_sold_address
+        ,CAST(t.contract_address as string) as project_contract_address
+        ,t.evt_tx_hash AS tx_hash
+        ,'' AS trace_address
+        ,t.evt_index
+    FROM
+        {{ source('uniswap_v3_arbitrum', 'Pair_evt_Swap') }} t
+    INNER JOIN 
+        {{ source('uniswap_v3_arbitrum', 'UniswapV3Factory_evt_PoolCreated') }} f
+        ON f.pool = t.contract_address
+    {% if is_incremental() %}
+    WHERE t.evt_block_time >= date_trunc("day", now() - interval '1 week')
+    {% endif %}
+)
+SELECT
+    'arbitrum' AS blockchain
+    ,'uniswap' AS project
+    ,'3' AS version
+    ,TRY_CAST(date_trunc('DAY', dexs.block_time) AS date) AS block_date
+    ,dexs.block_time
+    ,erc20a.symbol AS token_bought_symbol
+    ,erc20b.symbol AS token_sold_symbol
+    ,case
+        when lower(erc20a.symbol) > lower(erc20b.symbol) then concat(erc20b.symbol, '-', erc20a.symbol)
+        else concat(erc20a.symbol, '-', erc20b.symbol)
+    end as token_pair
+    ,dexs.token_bought_amount_raw / power(10, erc20a.decimals) AS token_bought_amount
+    ,dexs.token_sold_amount_raw / power(10, erc20b.decimals) AS token_sold_amount
+    ,dexs.token_bought_amount_raw
+    ,dexs.token_sold_amount_raw
+    ,coalesce(
+        dexs.amount_usd
+        ,(dexs.token_bought_amount_raw / power(10, p_bought.decimals)) * p_bought.price
+        ,(dexs.token_sold_amount_raw / power(10, p_sold.decimals)) * p_sold.price
+    ) AS amount_usd
+    ,dexs.token_bought_address
+    ,dexs.token_sold_address
+    ,coalesce(dexs.taker, tx.from) AS taker -- subqueries rely on this COALESCE to avoid redundant joins with the transactions table
+    ,dexs.maker
+    ,dexs.project_contract_address
+    ,dexs.tx_hash
+    ,tx.from AS tx_from
+    ,tx.to AS tx_to
+    ,dexs.trace_address
+    ,dexs.evt_index
+FROM dexs
+INNER JOIN 
+    {{ source('arbitrum', 'transactions') }} tx
+    ON tx.hash = dexs.tx_hash
+    {% if not is_incremental() %}
+    AND tx.block_time >= '{{project_start_date}}'
+    {% endif %}
+    {% if is_incremental() %}
+    AND tx.block_time >= date_trunc("day", now() - interval '1 week')
+    {% endif %}
+LEFT JOIN {{ ref('tokens_erc20') }} erc20a
+    ON erc20a.contract_address = dexs.token_bought_address 
+    AND erc20a.blockchain = 'arbitrum'
```

Comment:
there doesn't appear to be an erc20 tokens view for arbitrum:
```
select blockchain, count(1)
from tokens.erc20
group by blockchain
order by 1 desc
```

here is code for other blockchains:
https://github.com/duneanalytics/spellbook/tree/main/models/tokens

with this info missing, the join returns nothing, then all fields in the select above using either ```erc20a``` or ```erc20b``` return nulls.

example:
```
select
  *
from
  dbt_jeff_uniswap_v3_arbitrum.trades
where
  block_date >= '2022-10-10'
```

are you comfortable trying to add that model for arbitrum erc20 tokens?

Reply:
I would certainly be up to the challenge but definitely need a nudge in the right direction.  They are all hard coded so there must be a method to grabbing them.  Looking into Arbiscan's API now but not seeing anything obvious.

Reply:
Went ahead and started just pulling them manually.  Did not take as long as I expected.  I did the first 50 by market cap.

Reply:
thank you for the addition! i wanted to share one last thing that may help prior to moving towards the merge.

here is a query (stolen from another useful wizard) for tracking missing tokens in the view(s):
*note since arbitrum is starting from scratch, the missing token portion is commented out
https://dune.com/queries/1399945

if you'd like to take a look and consider adding any more of the most used tokens on uniswap v3 for arbitrum (this is just in the last week, you could open up further date wise), feel free to do so

Reply:
in a similar fashion, for most complete ```amount_usd``` values, this file is maintained in a similar fashion:
https://github.com/duneanalytics/spellbook/blob/main/models/prices/prices_tokens.sql

you can add the same subset of tokens to this file, it's a slightly different format but same idea

note that the ```token_id``` column is important for accuracy there, as it tells us how to call the coinpaprika API. that is where you can validate the id. here is helpful link:
https://api.coinpaprika.com/v1/coins/

addresses can be searched and id's found

you are the first to add arbitrum dex trades data, so it comes with a bit more work sadly 😅

Reply:
Fun! I had no idea I was headed down such a rabbit hole.  That's exciting though that I can help lay the foundation to keep building Arbi out.  Thanks for the query.  I reworked it a bit so it will only return addresses I had not already added.  Maybe someone will find it useful later -->[https://dune.com/queries/1401328].  
Will work on prices tomorrow.  Thanks for sharing the API.

Reply:
sure thing, look forward to seeing the add!